numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2365,9 +2363,166 @@ static PySequenceMethods voidtype_as_sequence = {
+{
+    if ((flags & PyBUF_WRITEABLE) == PyBUF_WRITEABLE) {
+        return -1;
+    }
+    Py@Name@ScalarObject *scalar = (Py@Name@ScalarObject *)self;
+
+    static char fmt[3] = "@fmt@";
+
+    view->ndim = 0;
+    view->len = sizeof(scalar->obval);
+    view->itemsize = sizeof(scalar->obval);
+    view->shape = NULL;
+    view->strides = NULL;
+    view->suboffsets = NULL;
+    Py_INCREF(self);
+    view->obj = self;
+    view->buf = &(scalar->obval);
+
+    if (((flags & PyBUF_FORMAT) != PyBUF_FORMAT)) {
+        /* It is unnecessary to find the correct format */
+        view->format = NULL;
+        return 0;
+    }
+
+    view->format = fmt;
+
+    return 0;
+}
+
+static PyBufferProcs @name@_arrtype_as_buffer = {
+    .bf_getbuffer = @name@_getbuffer,
+    /* No need to release the buffer */
+};
+
+/**end repeat**/
+
+static int
+unicode_getbuffer(PyObject *self, Py_buffer *view, int flags)
+{
+    if ((flags & PyBUF_WRITEABLE) == PyBUF_WRITEABLE) {
+        return -1;
+    }
+    PyUnicodeScalarObject *scalar = (PyUnicodeScalarObject *)self;
+    Py_ssize_t length = PyUnicode_GetLength(self);
+
+    view->ndim = 0;
+    view->len = length * 4;
+    view->itemsize = length * 4;
+    view->shape = NULL;
+    view->strides = NULL;
+    view->suboffsets = NULL;
+    Py_INCREF(self);
+    view->obj = self;
+
+    if (scalar->obval == NULL) {
+        /*
+         * Unicode may not have the representation available, `scalar_value`
+         * ensures materialization.
+         */
+        PyArray_Descr *descr = PyArray_DescrFromType(NPY_UNICODE);
+        scalar_value(self, descr);
+        Py_DECREF(descr);
+        if (scalar->obval == NULL) {
+            /* allocating memory failed */
+            Py_SETREF(view->obj, NULL);
+            return -1;
+        }
+    }
+    view->buf = scalar->obval;
+
+    if (((flags & PyBUF_FORMAT) != PyBUF_FORMAT)) {
+        /* It is unnecessary to find the correct format */
+        view->format = NULL;
+        return 0;
+    }
+
+    if (scalar->buffer_fmt != NULL) {
+        view->format = scalar->buffer_fmt;
+    }
+    else {
+        scalar->buffer_fmt = PyObject_Malloc(22);
+        if (scalar->buffer_fmt == NULL) {
+            Py_SETREF(view->obj, NULL);
+            return -1;
+        }
+        PyOS_snprintf(scalar->buffer_fmt, 22, "%" NPY_INTP_FMT "w", length);
+        view->format = scalar->buffer_fmt;
+    }
+
+    return 0;
+}
+
+static PyBufferProcs unicode_arrtype_as_buffer = {
+    .bf_getbuffer = unicode_getbuffer,
+    /* No need to release the buffer */
+};
+
+
+/**begin repeat
+ * #name = datetime, timedelta#
+ * #Name = Datetime, Timedelta#
+ */
+
+static int
+@name@_getbuffer(PyObject *self, Py_buffer *view, int flags)
+{
+    if ((flags & PyBUF_WRITEABLE) == PyBUF_WRITEABLE) {
+        return -1;
+    }
+    Py@Name@ScalarObject *scalar = (Py@Name@ScalarObject *)self;
+
+    view->ndim = 1;
+    view->len = 8;
+    view->itemsize = 1;
+    view->shape = &(view->len);
```

Comment:
Changes this to point at len internal to the struct instead of a global static...

Reply:
Is this safe? Are users allowed to memcpy this struct?

Reply:
Hmmm, interesting point. The user must release the `PyBuffer` struct specifically (after which the pointers become dead). And they are not allowed to e.g. mess with `view->obj` before that... So not sure if you can construct a weird usage that can break this, but happy to revert.

Reply:
My fear is use like:
```
PyBuffer b1;
{
    PyBuffer b2;
    get_buffer(my_obj, &b2);
    b1 = b2;
}
b1.shape; // is a dangling pointer!
```

Reply:
Yes, but that code at some point must call:
```
PyBuf_Release(b2)
```
and up until then, that pointer remains valid (unless the user overwrites `len` here). And after that, the pointer is invalid by definition.

Reply:
In my example and your implementation. b2.shape points at b1.length, but b1 no longer exists. So either your implementation is unsafe, or the line `b1 = b2` is forbidden by the buffer protocol.

Reply:
Well, changed it. It seems a bit funny, but may be technically allowed. The reason I find it a bit strange is that you will have to call `PyBuf_Release(b1)` even though you got the buffer initially as `b2` (otherwise its safe).

Reply:
Whether it's funny depends on whether _"The same bufferinfo struct should be used in the release-buffer interface call"_ from the pep requires only that `b1 == b2` (same value), or if it more strictly requires `&b1 == &b2` (same storage location).